File: leet_DP_0072_edit_distance.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        return [
            # DFS,
            BFS, # fastest
        ][0](word1, word2)

def BFS(word1: str, word2: str) -> int:
    len1 = len(word1)
    len2 = len(word2)
    Q = deque()
    Q.append((0,0))
    SEEN = set()
    res = 0
    while Q:
        N = len(Q)
        for _ in range(N):
            l, r = Q.popleft()
            # print(l,r)
            if (l, r) in SEEN:
                continue
            while l < len1 and r < len2 and word1[l] == word2[r]:
                l, r = l + 1, r + 1
            if l == len1 and r == len2:
                return res

            # insert in w1
            # advance index in w2 
            if (l, r + 1) not in SEEN:
                Q.append( (l, r + 1) )

            # replace an existing char in w1
            # advance both indexes
            if (l + 1, r + 1) not in SEEN:
                Q.append( (l + 1, r + 1) )

            # del char in w1
            if (l + 1, r) not in SEEN:
                Q.append( (l + 1, r) )
            SEEN.add( (l, r) )
        res += 1
    print('never here/')
    assert False

def DFS(word1: str, word2: str) -> int:
    len1 = len(word1)
    len2 = len(word2)

    # dp (memo)
    # l -> 0 - len1
    # r -> 0 - len2
    # dp of space -> len1 * len2
    hasCache = [[False] * (len2 + 1) for _ in range(len1 + 1)]
    cache = [[-1] * (len2 + 1) for _ in range(len1 + 1)]

    def f(l, r):
        if l == len1:
            return len2 - r
        if r == len2:
            return len1 - l
        if hasCache[l][r]:
            return cache[l][r]
        res = 1e20
        if word1[l] == word2[r]:
            res = min(res, f(l + 1, r + 1))
        res = min(res, f(l, r + 1) + 1) # insert in w1 ie. delete from w2
        res = min(res, f(l + 1, r) + 1) # delete from w1
        res = min(res, f(l + 1, r + 1) + 1) # replace

        hasCache[l][r] = True
        cache[l][r] = res
        return res

    return f(0, 0)

File: leet_DP_0072_edit_distance.cc
class Solution {
public:
    int minDistance(string word1, string word2)
    {
        vector<int(*)(string, string)> Solutions = {
            // dfs,
            bfs,
        };
        return Solutions[0](word1, word2);
    }

    static int bfs (string w1, string w2)
    {
        int n1 = w1.length();
        int n2 = w2.length();
        deque<vector<int>> Q{ {0, 0} };
        set<vector<int>> SEEN;
        int res = 0;
        while (!Q.empty())
        {
            int N = Q.size();
            int i = -1;
            while (++i < N)
            {
                auto top = Q.front();
                Q.pop_front();
                int l = top[0], r = top[1];
                if (SEEN.find({l, r}) != end(SEEN))
                    continue ;
                
                while (l < n1 && r < n2 && w1[l] == w2[r])
                {
                    ++l;
                    ++r;
                }
                if (l == n1 && r == n2)
                    return res;

                vector<int> p;

                // INSERT
                p = {l, r + 1};
                if (SEEN.find(p) == end(SEEN))
                    Q.push_back(p);
                // REPLACE
                p = {l + 1, r + 1};
                if (SEEN.find(p) == end(SEEN))
                    Q.push_back(p);
                // DELETE
                p = {l + 1, r};
                if (SEEN.find(p) == end(SEEN))
                    Q.push_back(p);
                SEEN.insert( {l, r} );
            }
            res++;
        }
        assert (false);
        return -1;
    }

    static int dfs (string w1, string w2)
    {
        int n1 = w1.length();
        int n2 = w2.length();
        vector<vector<bool>> cached( n1 + 1, vector<bool>(n2 + 1, false) );
        vector<vector<int>> caching( n1 + 1, vector<int>(n2 + 1, -1) );
        function<int(int, int)> f/*(int, int)*/ = [&](int l, int r) -> int {
            if (l == n1)
                return n2 - r;
            if (r == n2)
                return n1 - l;
            if (cached[l][r])
                return caching[l][r];
            int res = INT_MAX;
            int temp;
            if (w1[l] == w2[r])
                res = min(f(l + 1, r + 1), res);
            // INSERT
            res = min(f(l, r + 1) + 1, res);
            // DELETE
            res = min(f(l + 1, r) + 1, res);
            // REPLACE
            res = min(f(l + 1, r + 1) + 1, res);
            cached[l][r] = true;
            caching[l][r] = res;
            return res;
        };
        return f (0, 0);
    }
};
